File: services/auth-service/src/main/java/com/dennis/ecommerce/auth/config/RabbitMQConfig.java
package com.dennis.ecommerce.auth.config;

import org.springframework.amqp.core.*;
import org.springframework.amqp.rabbit.connection.ConnectionFactory;
import org.springframework.amqp.rabbit.core.RabbitTemplate;
import org.springframework.amqp.support.converter.Jackson2JsonMessageConverter;
import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;

@Configuration
public class RabbitMQConfig {

    public static final String USER_EXCHANGE = "user.exchange";
    public static final String USER_CREATED_QUEUE = "user.created.queue";
    public static final String USER_CREATED_ROUTING_KEY = "user.created";

    @Bean
    public TopicExchange userExchange() {
        return new TopicExchange(USER_EXCHANGE);
    }

    @Bean
    public Queue userCreatedQueue() {
        return QueueBuilder.durable(USER_CREATED_QUEUE).build();
    }

    @Bean
    public Binding userCreatedBinding() {
        return BindingBuilder
                .bind(userCreatedQueue())
                .to(userExchange())
                .with(USER_CREATED_ROUTING_KEY);
    }

    @Bean
    public Jackson2JsonMessageConverter messageConverter() {
        return new Jackson2JsonMessageConverter();
    }

    @Bean
    public RabbitTemplate rabbitTemplate(ConnectionFactory connectionFactory) {
        RabbitTemplate template = new RabbitTemplate(connectionFactory);
        template.setMessageConverter(messageConverter());
        return template;
    }
}

File: services/catalog-service/src/main/kotlin/com/dennis/ecommerce/catalogService/config/RabbitMQConfig.kt
package com.dennis.ecommerce.catalogService.config

import org.springframework.amqp.core.*
import org.springframework.amqp.rabbit.connection.ConnectionFactory
import org.springframework.amqp.rabbit.core.RabbitTemplate
import org.springframework.amqp.support.converter.Jackson2JsonMessageConverter
import org.springframework.amqp.support.converter.MessageConverter
import org.springframework.beans.factory.annotation.Value
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration

@Configuration
class RabbitMQConfig {

    // ── Exchanges ─────────────────────────────────────────
    @Value("\${messaging.exchange.catalog}")
    private lateinit var catalogExchange: String

    @Value("\${messaging.exchange.order}")
    private lateinit var orderExchange: String

    // ── Queues ────────────────────────────────────────────
    @Value("\${messaging.queues.order-confirmed}")
    private lateinit var orderConfirmedQueue: String

    @Value("\${messaging.queues.order-cancelled}")
    private lateinit var orderCancelledQueue: String

    // ── Routing Keys ──────────────────────────────────────
    @Value("\${messaging.routing-keys.order-confirmed}")
    private lateinit var orderConfirmedRoutingKey: String

    @Value("\${messaging.routing-keys.order-cancelled}")
    private lateinit var orderCancelledRoutingKey: String

    // ── Exchanges ─────────────────────────────────────────

    // Exchange propio del catalog-service
    @Bean
    fun catalogExchange(): TopicExchange =
        TopicExchange(catalogExchange, true, false)

    // Exchange del order-service (solo lo referenciamos, no lo creamos porque ya debe haber sido cerado por el order-service cuando p[ubliquen los mensajes que leeremos aqui)
    @Bean
    fun orderExchange(): TopicExchange =
        TopicExchange(orderExchange, true, false)

    // ── Queues ────────────────────────────────────────────

    @Bean
    fun orderConfirmedQueue(): Queue =
        QueueBuilder.durable(orderConfirmedQueue)
            .withArgument("x-dead-letter-exchange", "$catalogExchange.dlx")
            .build()

    @Bean
    fun orderCancelledQueue(): Queue =
        QueueBuilder.durable(orderCancelledQueue)
            .withArgument("x-dead-letter-exchange", "$catalogExchange.dlx")
            .build()

    // ── Bindings ──────────────────────────────────────────

    // Escucha eventos del order-service
    @Bean
    fun orderConfirmedBinding(): Binding =
        BindingBuilder
            .bind(orderConfirmedQueue())
            .to(orderExchange())
            .with(orderConfirmedRoutingKey)

    @Bean
    fun orderCancelledBinding(): Binding =
        BindingBuilder
            .bind(orderCancelledQueue())
            .to(orderExchange())
            .with(orderCancelledRoutingKey)

    // ── Serialización JSON ────────────────────────────────

    @Bean
    fun jsonMessageConverter(): MessageConverter =
        Jackson2JsonMessageConverter()

    @Bean
    fun rabbitTemplate(connectionFactory: ConnectionFactory): RabbitTemplate =
        RabbitTemplate(connectionFactory).apply {
            messageConverter = jsonMessageConverter()
        }
}